documentation/general/dotnet-run-file.md

```diff
@@ -0,0 +1,224 @@
+(an error is reported if it points to a directory instead).
+Then the build ignores other entry-point files.
+
+Thanks to this, it is possible to have a structure like
+```
+App/Util.cs
+App/Program1.cs
+App/Program2.cs
+```
+where either `Program1.cs` or `Program2.cs` can be run and both of them have access to `Util.cs`.
+
+In this case, there are multiple implicit projects (and similarly multiple projects are materialized during [grow up](#grow-up)):
+```
+App/Shared/Util.cs
+App/Shared/Shared.csproj
+App/Program1/Program1.cs
+App/Program1/Program1.csproj
+App/Program2/Program2.cs
+App/Program2/Program2.csproj
+```
+
+The generated folders might need to be named differently to avoid clashes with existing folders.
+
+Unless the [artifacts output layout][artifacts-output] is used (which is recommended),
+those implicit projects mean that build artifacts are placed under those implicit directories
+even though they don't exist on disk prior to build:
+```
+App/Shared/bin/
+App/Shared/obj/
+App/Program1/bin/
+App/Program1/obj/
+App/Program2/bin/
+App/Program2/obj/
+```
+
+## Package references
+
+It is possible to specify NuGet package references via the `#r` directive.
+
+```cs
+#r "nuget: Newtonsoft.Json, 13.0.1"
+```
+
+The C# language needs to be updated to ignore these directives (instead of failing the compilation).
+See [the corresponding language proposal][pound].
+
+If these directives were limited by the language to only appear near the top of the file (similar to `#define` directives),
+the dotnet CLI could be more efficient in searching for them.
+
+We can also consider reporting an error or warning if `#r` directives are spread among multiple files,
+to encourage users to keep them in a single place for clarity.
+However, we cannot merely allow them just in entry-point files,
+because that would mean non-entry-point files in [multi-entry-point scenarios](#multiple-entry-points)
+could not reference external NuGet packages.
+
+## Shebang
+
+Along with `#r`, the language can also ignore `#!` which could be then used for [shebang][shebang] support.
+
+```cs
+#!/usr/bin/dotnet run
+Console.WriteLine("Hello");
+```
+
+It might be beneficial to also ship `dotnet-run` binary
+because some shells do not support multiple command-line arguments in the shebang
+which is needed if one wants to use `/usr/bin/env` to find the `dotnet` executable
+(although `-S` argument can be sometimes used to enable multiple argument support):
+
+```cs
+#!/usr/bin/env dotnet run
+// ^ Might not work in all shells. "dotnet run" might be passed as a single argument to "env".
+```
+```cs
+#!/usr/bin/env dotnet-run
+// ^ Should work in all shells.
+```
+```cs
+#!/usr/bin/env -S dotnet run
+// ^ Workaround in some shells.
+```
+
+## Other commands
+
+We can consider supporting other commands like `dotnet build`, `dotnet pack`, `dotnet watch`.
+
+## Implementation
+
+The build is performed using MSBuild APIs on in-memory project files.
+
+### Optimizations
```

Comment:
For running a single file in a larger directory with lots of files, is the expectation that we are parsing all the files, figuring out the right set inputs, and then reparsing all a second time to actually run this?

Reply:
If we allowed just a single entry point and limited the `#r`s to only appear there, we could just look at that single file. If we allow multiple entry points (and hence need to allow `#r`s everywhere so non-entry-point files can reference NuGet packages as well), we need to look at all input files.

But we don't need to parse them completely, we can just quickly find `#r`s via a regex.

I guess we could also say `#r`s can only appear in a single file per virtual project, i.e., in each entry-point file and one non-entry-point file. We can then only look at that one file for incremental re-runs.